src/resources/uart_esp32.cc

```diff
@@ -17,95 +17,584 @@
+}
+
+void UartResource::enable_read_interrupts() {
+  enable_interrupt_index(UART_TOIT_INTR_RXFIFO_FULL);
+  enable_interrupt_index(UART_TOIT_INTR_RX_TIMEOUT);
+}
+
+void UartResource::disable_interrupt_index(uart_toit_interrupt_index_t index) {
+  SpinLocker locker(&spinlock_);
+  disable_interrupt_mask_(interrupt_mask(index));
+}
+
+void UartResource::clear_interrupt_index(uart_toit_interrupt_index_t index) {
+  uart_toit_hal_clr_intsts_mask(hal_, interrupt_mask(index));
+}
+
+void UartResource::clear_interrupt_mask(uint32_t mask) {
+  uart_toit_hal_clr_intsts_mask(hal_, mask);
+}
+
+bool UartResource::try_set_iomux_pin(gpio_num_t pin, uint32_t iomux_index) const {
+  const uart_periph_sig_t *uart_pin = &uart_periph_signal[port_].pins[iomux_index];
+  if (uart_pin->default_gpio == -1 || uart_pin->default_gpio != pin) return false;
+  gpio_iomux_out(pin, uart_pin->iomux_func, false);
+  if (uart_pin->input) gpio_iomux_in(pin, uart_pin->signal);
+  return true;
+}
+
+size_t UART_ISR_INLINE UartResource::get_tx_fifo_available_count() {
+  return uart_toit_hal_get_txfifo_len(hal_);
+}
+
+UART_ISR_INLINE uart_event_types_t UartResource::interrupt_handler_read() {
+  size_t read_length = uart_toit_hal_get_rxfifo_len(hal_);
+  if (read_length == 0) return UART_EVENT_MAX;
+
+  size_t rx_buffer_free = rx_buffer_.free_count();
+  if (rx_buffer_free == 0) {
+    disable_read_interrupts();
+    return UART_BUFFER_FULL;
+  } else {
+    if (rx_buffer_free < read_length) read_length = rx_buffer_free;
+    uint8_t buffer[read_length];
+    uart_toit_hal_read_rxfifo(hal_, buffer, reinterpret_cast<int*>(&read_length));
+    rx_buffer().send(buffer, read_length);
+    return UART_DATA;
+  }
+}
+
+UART_ISR_INLINE uart_event_types_t UartResource::interrupt_handler_write() {
+  uint32_t tx_fifo_free = get_tx_fifo_available_count();
+  size_t received_count;
+  uint8_t* buffer = tx_buffer().read(&received_count, tx_fifo_free);
+  if (buffer) {
+    uint32_t written;
+    uart_toit_hal_write_txfifo(hal_, buffer, received_count, &written); // TODO: Is the RS485 workaround necessary here?
+    tx_buffer().return_buffer(buffer);
+    return UART_TX_EVENT;
+  } else {
+    int break_number = tx_buffer().read_break();
+    if (break_number != 0) {
+      enable_interrupt_index_isr(UART_TOIT_INTR_TX_BRK_DONE);
+      uart_toit_hal_tx_break(hal_, break_number);
+      disable_interrupt_index_isr(UART_TOIT_INTR_TXFIFO_EMPTY);
+    } else if (tx_buffer().is_empty()) {
+      disable_interrupt_index_isr(UART_TOIT_INTR_TXFIFO_EMPTY);
+    }
+    return UART_EVENT_MAX;
+  }
+}
+
+IRAM_ATTR void UartResource::interrupt_handler(void* arg) {
+  auto uart = unvoid_cast<UartResource*>(arg);
+  uint32_t uart_interrupt_status = uart_toit_hal_get_intsts_mask(uart->hal_);
+  if (uart_interrupt_status == 0) return;
+  uart_event_types_t event = UART_EVENT_MAX;
+
+  if (uart_interrupt_status & uart->interrupt_mask(UART_TOIT_INTR_TXFIFO_EMPTY)) {
```

Comment:
I think the calls to interrupt_mask will return the same every time, and perhaps at some point (not now) we should cache them to improve the performance of this dispatcher.

Are we sure that only one can be true?  Should we loop until all bits have been handled?

Reply:
Multiple interrupts can be set at once. However, since we do not clear the interrupt flag, the handle will just be called multiple times. I found it more readable this way. It might be interesting to explore what is faster.